File: codes/Tree/112.path-sum/path-sum.cpp
#include <queue>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {
    }
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {
    }
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {
    }
};

class Solution {
public:
    int targetSum;
    bool hasPathSum(TreeNode *root, int targetSum) {
        if (!root) {
            return false;
        }
        this->targetSum = targetSum;
        return preOrder(root,0);
    }
    bool preOrder(TreeNode *node, int cur) {
        cur += node->val;
        if (!node->left && !node->right && cur == targetSum) {
            return true;
        }
        bool result = false;

        if (node->left) {
            result |= preOrder(node->left, cur);
        }
        if (node->right) {
            result |= preOrder(node->right, cur);
        }
        return result;
    }
};

File: codes/Tree/112.path-sum/path-sum.go
package main
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}


func hasPathSum(root *TreeNode, targetSum int) bool {
    if root==nil{
        return false
    }
    return dfs(root,targetSum)
}
func dfs(root* TreeNode,targetSum int)bool{
    if root==nil{
        return false
    }
    targetSum-=root.Val
    if root.Left==nil && root.Right==nil&&targetSum==0{
        return true
    }
    return dfs(root.Left,targetSum)||dfs(root.Right,targetSum)
}